### DISCUSSION TITLE
Scalar bar doesn't update after plotter.render() or plotter.add()

### DISCUSSION THREAD

**Discussion by jsaintauret** [marcomusy/vedo#655] [Q&A]
Hi

when I update the scalar attach to my mesh with the following the scalar bar isn't updating:
```
mesh.pointdata["myscalars1"] = new_scalar
mesh.cmap(my_colormap, 'myscalars1')
mesh.addScalarBar()
plotter.render() or  plotter.remove('my_mesh').add(mesh)
```
the only thing working is if I first clear the plotter with` plotter.clear() `and then add back the mesh with show()

it's weird because when I go into the debug mode an check the TableRange of  the  scalarbar attached to the  mesh the range is -950,850 as expected:
![image](https://user-images.githubusercontent.com/52704901/170671036-5af4a048-d803-46b4-a942-71b2ee6c8269.png)
 
But the TableRange of the scalarbar attached to the plotter instance is still the previous range -950,2113 (it means that the scalarbrs attribute of the plotter has not been updated after the add function)

![image](https://user-images.githubusercontent.com/52704901/170672374-60be8c47-8cc3-432b-8aba-8cff6274a249.png)

So should the plotter updates all the actor to show when I call render() ?or when I use add(mesh) I expect the plotter/render to update all mesh properties accordingly

**Comment by marcomusy** [comment]
hi, if the mesh data is modified there no automatism in place to update the scalarbar (there should be, sorry about that).
the easiest way out is either 
- remove it completely (the information is redundant anyway)
- remove the scalarbar with `.remove(mymesh.scalarbar)` and then recreate it with `.addScalarBar()` (you may need to also `plotter.add(mymesh.scalarbar, render=False)`)

**Comment by jsaintauret** [comment]
Thanks perfect! It works
![capture (4)](https://user-images.githubusercontent.com/52704901/170717246-c879d56a-c45a-4edf-bec1-1c4c3cbaa7ad.gif)

One quick question, any reason why the 0 is dropped from the axis label ? any workaround I want to show it ? (check the end of the GIF to see what I'm talking about)

**Reply by marcomusy** [reply]
> any reason why the 0 is dropped from the axis label ?

uhm not sure why that would happen sorry :(

**Reply by jsaintauret** [reply]
Ok no problem , I'll debug and see what's going on and let you know

**Reply by jsaintauret** [reply]
Hey,

I think I've found the issue! From what I see here it looks like if the label return by `zticks_str[i]`  is equal to 0 and so the code continue to the next iteration therefore skipped any label equal to 0

![image](https://user-images.githubusercontent.com/52704901/171204274-51820add-5956-443f-94cb-0d4e3a148126.png)

I don't know if it's intended or not but maybe replace by `if t is None` or something similar to check if that specific label is empty would avoid skipping any label equal to 0?

**Reply by marcomusy** [reply]
where is this code?

**Reply by jsaintauret** [reply]
Sorry, it is addon.py/Axes class 
![image](https://user-images.githubusercontent.com/52704901/171254294-1c8fad7e-28e4-4a48-9449-16cfab1b78c1.jpeg)

For each axis x y and z

**Reply by marcomusy** [reply]
ok, but I don't think the problem is there because in python `bool("0") is True`, while `bool("") is False`

**Reply by jsaintauret** [reply]
Yes I agree with you if the type is str, but in that case the type of the items in ticks_list is float

**Reply by jsaintauret** [reply]
Hey hello marco hope you are well.
I didn't not get an answer from you following what I said about the fact that the the items inside the list tick_list are floats.

One more thing that I wanted to ask is about updating the scalar attached to my mesh.
let say that I have this `mesh.pointdata["my_new_array"] = new_sets_of_scalar`, when I add the scalar bar I still get the range data of my mesh. Is there a way to update the scalar of my mesh or do I need to manually set the table range of the vtkLookupTable attached to the scalarbar?

Thanks

**Reply by marcomusy** [reply]
hi! the question above is not very clear to me, I cannot see the mistake in the source code.

second question
there is no automatic adjusting of the scalarbar so i'm afraid you need to recreate it if you change the pointdata content without changing the array name.
I have not checked but you may need to set
`mymesh.scalarbar = None`
then again:
`mymesh.addScalarBar()`
or set the `vtkLookupTable.SetRange()` manually as you suggested

**Comment by jsaintauret** [comment]
So here when I debug the code I see this 
![image](https://user-images.githubusercontent.com/52704901/183722903-df80b74d-4cc2-4a95-8714-bc1b445c929f.png)

the `ztick_str `list that holds my custom z labels appears to have items of type float, so in case one of the item is 0.0 (the index 2 of that list in my case) the code will enter in to that if block `if not t:` and thus continue and skip to display that value along that axis.

Hope my explanation is clear.

As for my last question , I do remember you telling that I should manually remove and add the scalar back to the plotter in order to simulate an automatic update, which I did. But what I'm confuse about is maybe the fact that even though I update the scalar attached to the pointdata prior to adding the scalar bar then my scalar range doesn't get update  on the scalar bar afterward.
So yes  I'm going to try to do it manually with `vtkLookupTable.SetRange()`, but could please explain your point here:

>i'm afraid you need to recreate it if you change the pointdata content without changing the array name.
Is there a default array name ? the original array?

**Reply by marcomusy** [reply]
Oh now I see what the problem can be... you are passing the axes labels manually using `xValuesAndLabels`, but you should specify them as strings not floats.
I will force them to be strings in the next release.
About the second question, I'll try to work out  a test example to reproduce the issue.

**Comment by jsaintauret** [comment]
Cool, thanks indeed I need to convert my list element into string, I will do that.

Thanks for the example I will wait then.

**Reply by marcomusy** [reply]
can you reproduce this
```python
from vedo import *
s = Sphere().lw(0.1)
pts = s.points()

plt = Plotter(axes=1)

scalars = pts[:,2]
s.cmap("winter", scalars).addScalarBar() # or addScalarBar3D()
plt.show(s)

plt.remove(s.scalarbar)

scalars2 = pts[:,2] *2
s.cmap("winter", scalars2).addScalarBar()
plt.show(s)
plt.close()
```
![Screenshot from 2022-08-10 12-24-21](https://user-images.githubusercontent.com/32848391/183879111-fcf56431-9a78-4ac9-b1dd-a71eb14cf8cd.png)

**Comment by jsaintauret** [comment]
Perfect! I have the label 0 on Z axe and the scalar up to date in the scalarbar 
![image](https://user-images.githubusercontent.com/52704901/183895312-1a9a65b2-ce7e-405f-95f3-626d69c55b61.png)

**Reply by marcomusy** [reply]
Cool :)
you can also try `Axes(..., yLabelJustify="center-left")` to avoid label overlapping

![Screenshot from 2022-08-10 17-53-59](https://user-images.githubusercontent.com/32848391/183954744-8325c2a6-eb8d-4f2b-ac03-ed79aa718fb0.png)

**Reply by jsaintauret** [reply]
Ok I will do it!

I've been also meaning to ask you something about vedo. 

So my goal is to build a visualization tool based on vedo for a specific application in chemistry, I understand that vedo was not aimed to be used as a "**dynamic**" visualization tool but so far I was able to work something out for my need although I'm still thinking the tool needs to improve perfomance wise: 

Here is a quick overview of the tool

![image](https://user-images.githubusercontent.com/52704901/183973345-f58ebada-4279-4bed-8da0-5bfe92e8ceb2.png)


One of the feature of this tool (which is a basic feature for visualization tools) is to be able to select  the axes limit in order to zoom in some area of the mesh. To do that I clip the `panda.dataframe` dataset according to a specific range and then I update the polydatas of the mesh accordingly and finaly I scale up/down the mesh so it can fit the current (initial) axes bounding.

I saw that a `cutWithBox` method exist that could do the trick (minus the scaling/fit part) but I was wondering if I put back the initial bounding box value to restore the mesh view after the cut, would the initial  mesh bounding be back?

**Reply by marcomusy** [reply]
> I understand that vedo was not aimed to be used as a "dynamic" visualization tool 

well ... that depends on how much dynamic your tool is meant to be :)

> I saw that a `cutWithBox` method exist that could do the trick (minus the scaling/fit part) but I was wondering if I put back the initial bounding box value to restore the mesh view after the cut, would the initial mesh bounding be back?

what I would do here is to keep the initial uncut mesh and then create a copy of it on the fly with `mesh.clone(deep=False)` on which you apply all cuts  with `cutWithBox` and customizations.
when you change ranges or anything else you throw the copy away and create a new one.

**Reply by jsaintauret** [reply]
> well ... that depends on how much dynamic your tool is meant to be :)

for the first iteration of the tool lol, I would say: mesh range update, color changing /size/oriention of the axe label, switch the mesh type from surface to contour mesh ect. Realy basic stuff 

> what I would do here is to keep the initial uncut mesh and then create a copy of it on the fly with `mesh.clone(deep=False)` on which you apply all cuts with `cutWithBox` and customizations. when you change ranges or anything else you throw the copy away and create a new one.

Ok I get the idea, so perform every customization on a clone then throw it away when I want to change the customazation and so on
